FastAPI Project setup

boghtml opened the pull request:
dev
## JIRA

* [Main JIRA ticket](https://jira.softserve.academy/secure/RapidBoard.jspa?rapidView=id)


## Code reviewers

- [ ] @github_username

### Second Level Review

- [ ] @github_username

## Summary of issue

ToDo

## Summary of change

ToDo

## Testing approach

ToDo

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

## Made:
- launched the project
- connected Docker
- added models and migrations

### For it to work, you need to add the following `.env` file to the project directory:

```env
# .env
APP_NAME="Forum MicroService"
DEBUG=True

# Cassandra settings
CASSANDRA_CONTACT_POINTS=["cassandra"]
CASSANDRA_KEYSPACE=forum_keyspace
CASSANDRA_PORT=9042
CASSANDRA_USERNAME=cassandra
CASSANDRA_PASSWORD=cassandra

APP_HOST=0.0.0.0
APP_PORT=8000

CORS_ORIGINS=http://localhost

CASSANDRA_MAX_RETRIES=5
CASSANDRA_RETRY_DELAY=5

IraMladanovych02 commented on BackEnd/forum-service/sn-forum-service/app/schemas/post.py:
```diff
@@ -0,0 +1,36 @@
+    post_id: int
+    title: str
+    author_id: int
+    created_at: datetime
```
If created_at is always set to "now" at creation, you can use add created_at: datetime = Field(default_factory=datetime.utcnow)

boghtml replied:
added

IraMladanovych02 commented on BackEnd/forum-service/sn-forum-service/app/schemas/post.py:
```diff
@@ -0,0 +1,36 @@
+    content: str
+    categories: Optional[List[Dict[str,str]]] = []
+    tags: Optional[List[Dict[str,str]]] = []
+    views_count: int
```
if when creating a post views_count and likes_count always starts with 0, you can add views_count: int = 0
likes_count: int = 0

boghtml replied:
added

IraMladanovych02 commented on BackEnd/forum-service/sn-forum-service/app/schemas/post.py:
```diff
@@ -0,0 +1,36 @@
+
+class PostCreate(BaseModel):
+    author_id: int
+    title: str
```
(optional) add limits to title field  - title: constr(max_length=200)

boghtml replied:
Added, thanks

IraMladanovych02 commented on BackEnd/forum-service/sn-forum-service/app/routes/questions.py:
```diff
@@ -0,0 +1,8 @@
+# app/routes/posts.py
+from fastapi import APIRouter
+
+router = APIRouter()
```
(optional) you can group your endpoints for documentation and using router = APIRouter(
    prefix="/posts",
    tags=["Posts"] 
)

IraMladanovych02 commented on BackEnd/forum-service/sn-forum-service/app/routes/posts.py:
```diff
@@ -0,0 +1,40 @@
+import logging
+
+logger = logging.getLogger(__name__)
+router = APIRouter()
```
(optional) - you can group endpoints router = APIRouter(
    prefix="/posts",
    tags=["Posts"]
)

boghtml replied:
Added, thanks

IraMladanovych02 commented on BackEnd/forum-service/sn-forum-service/app/routes/posts.py:
```diff
@@ -0,0 +1,40 @@
+    except Exception as e:
+        logger.error(f"Error fetching posts: {str(e)}")
+        raise HTTPException(
+            status_code=500,
```
please add status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,

IraMladanovych02 reviewed (changes requested):
Well done! added comments, please check it

mehalyna commented on BackEnd/forum-service/sn-forum-service/app/services/cassandra.py:
```diff
@@ -0,0 +1,55 @@
+            
+            session.set_keyspace(settings.cassandra_keyspace)
+            
+            cluster.register_user_type(settings.cassandra_keyspace, "category_info", dict)
+            cluster.register_user_type(settings.cassandra_keyspace, "tag_info", dict)
```
If Cassandra schema is not fully initialized, `register_user_type()` can fail.
Move `register_user_type()` to a separate function and call it after schema creation.

mehalyna commented on BackEnd/forum-service/sn-forum-service/docker-compose.yml:
```diff
@@ -0,0 +1,49 @@
+      - CASSANDRA_PASSWORD=cassandra
+    volumes:
+      - cassandra-data:/var/lib/cassandra
+    healthcheck:
+      test: ["CMD-SHELL", "cqlsh -u cassandra -p cassandra -e 'describe keyspaces'"]
+      interval: 15s
+      timeout: 10s
+      retries: 10
```
The health check only verifies keyspaces but doesn't check if Cassandra is fully ready. `cqlsh` may return results even before the database is fully initialized.
Use a more robust health check that waits until Cassandra accepts connections.
Example:
```
healthcheck:
  test: ["CMD", "cqlsh", "-u", "cassandra", "-p", "cassandra", "-e", "SELECT release_version FROM system.local"]
  interval: 15s
  timeout: 10s
  retries: 10
```

boghtml replied:
corrected

mehalyna commented on BackEnd/forum-service/sn-forum-service/docker-compose.yml:
```diff
@@ -0,0 +1,49 @@
+  cassandra:
+    image: cassandra:latest
+    container_name: cassandra_forum
+    ports:
```
Your code expects `CASSANDRA_PORT=9042`, but you expose 9062 instead.

boghtml replied:
I also left the external port as 9062 - because if I set it to 9042, I will have problems with ScalaDb in another project, since it runs on 9042:9042, so for now I left the port here as 9062:9042

mehalyna commented on BackEnd/forum-service/sn-forum-service/docker-compose.yml:
```diff
@@ -0,0 +1,49 @@
+    ports:
+      - "9062:9042"
+    environment:
+      - CASSANDRA_START_RPC=true
```
Looks like Cassandra 4.x+ does not use `CASSANDRA_START_RPC` anymore.

boghtml replied:
added

mehalyna commented on BackEnd/forum-service/sn-forum-service/docker-compose.yml:
```diff
@@ -0,0 +1,49 @@
+      options:
+        max-size: "10m"
+        max-file: "3"
+    depends_on:
+      cassandra:
+        condition: service_healthy
```
Even with `depends_on: condition: service_healthy`, Cassandra might still not be ready. The FastAPI app might start too early and fail to connect.
Add a startup delay using `entrypoint.sh`.
Example `entrypoint.sh`
```
#!/bin/sh

echo "Waiting for Cassandra to be ready..."
until cqlsh cassandra -u cassandra -p cassandra -e "SELECT release_version FROM system.local" > /dev/null 2>&1; do
  echo "Cassandra is not ready yet, waiting..."
  sleep 5
done

echo "Cassandra is ready, starting the application..."
exec uvicorn app.main:app --host 0.0.0.0 --port 8000
```
Modify `docker-compose.yml`
```
    entrypoint: ["sh", "./entrypoint.sh"]
```
Now, FastAPI waits until Cassandra is fully ready before starting.

boghtml replied:
added

vladmorozov2 replied:
Good job!

boghtml commented on BackEnd/forum-service/sn-forum-service/Dockerfile:
```diff
@@ -0,0 +1,15 @@
+RUN pip install --no-cache-dir --upgrade pip && \
+    pip install --no-cache-dir -r requirements.txt
+
+COPY . .
```
corrected

boghtml commented on BackEnd/forum-service/sn-forum-service/app/config.py:
```diff
@@ -0,0 +1,23 @@
+
+    class Config:
+        env_file = ".env"
+        case_sensitive = False
```
corrected

boghtml commented on BackEnd/forum-service/sn-forum-service/app/main.py:
```diff
@@ -0,0 +1,45 @@
+async def startup_event():
+    print("init casandra")
+
+    init_cassandra() 
```
corrected

boghtml commented:
- Added .dockerignore to exclude unnecessary files and improve build performance.
- Updated Dockerfile to install cqlsh for Cassandra readiness checks.
- Configured docker-compose.yml: aligned Cassandra ports (9062:9042), added entrypoint.sh for startup delay.
- Refactored cassandra.py: renamed to async_init_cassandra, added asyncio.Lock for thread safety, used cassandra_contact_points, added DCAwareRoundRobinPolicy.
- Updated config.py: added retry settings, fixed cassandra_contact_points parsing.
- Made migrations.py and seed_data.py async: used await get_session() for proper session handling.
- Adjusted main.py: updated imports, ensured async calls with await, added dynamic CORS, moved router registration to a function.
- Fixed .env: removed brackets from CASSANDRA_CONTACT_POINTS, aligned CASSANDRA_PORT with Cassandra's internal port (9042).
- Replaced print() with logger.info() across files for consistent logging.
- Resolved various runtime errors (ImportError, AttributeError, TypeError) related to async usage and configuration.

boghtml commented on BackEnd/forum-service/sn-forum-service/app/config.py:
```diff
@@ -0,0 +1,23 @@
+    app_name: str = Field("Forum MicroService", env="APP_NAME")
+    debug: bool = Field(True, env="DEBUG")
+    
+    cassandra_contact_points: list = Field(["cassandra"], env="CASSANDRA_CONTACT_POINTS")
```
corrected

boghtml commented on BackEnd/forum-service/sn-forum-service/app/config.py:
corrected

boghtml commented on BackEnd/forum-service/sn-forum-service/app/main.py:
corrected

IraMladanovych02 reviewed (approved):
good job!

vladmorozov2 reviewed (commented):
Great job!
